pretixprint/app/src/main/java/eu/pretix/pretixprint/connections/Sunmi.kt

```diff
@@ -0,0 +1,113 @@
+        val dpi = Integer.valueOf(conf.get("hardware_${type}printer_dpi")
+                ?: proto.defaultDPI.toString()).toFloat()
+
+        try {
+            Log.i("PrintService", "Starting renderPages")
+            val futures = renderPages(proto, tmpfile, dpi, numPages, conf, type)
+
+            Log.i("PrintService", "bindService")
+            InnerPrinterManager.getInstance().bindService(context, object : InnerPrinterCallback() {
+                override fun onConnected(printerService: SunmiPrinterService) {
+                    Log.i("PrintService", "PrinterService connected")
+                    when (proto) {
+                        is StreamByteProtocol<*> -> {
+                            proto.send(futures, bais, baos, conf, type)
+                            printerService.sendRAWData(baos.toByteArray(), object : InnerResultCallback() {
+                                override fun onRunResult(p0: Boolean) {
+                                    Log.i("PrintService", "PrinterService onRunResult: $p0")
+                                    future.complete(null)
+                                }
+
+                                override fun onReturnString(p0: String?) {
+                                    Log.i("PrintService", "PrinterService onReturnString: $p0")
+                                }
+
+                                override fun onRaiseException(code: Int, msg: String?) {
+                                    future.completeExceptionally(Exception("[$code] $msg"))
+                                }
+
+                                override fun onPrintResult(p0: Int, p1: String?) {
+                                    Log.i("PrintService", "PrinterService onPrintResult: $p0 $p1")
+                                }
+
+                            })
+                        }
+
+                        is SunmiByteProtocol<*> -> {
+                            proto.sendSunmi(printerService, futures, conf, type)
+                            future.complete(null)
+                        }
+
+                        is CustomByteProtocol<*> -> {
+                            throw RuntimeException("Combination not supported")
+                        }
+                    }
+                }
+
+                override fun onDisconnected() {
+                    Log.i("PrintService", "PrinterService onDisconnected")
+                }
+            })
+        } catch (e: TimeoutException) {
+            e.printStackTrace()
+            throw PrintException("Rendering timeout, thread may have crashed")
+        } catch (e: PrintError) {
+            e.printStackTrace()
+            throw PrintException(context.applicationContext.getString(R.string.err_job_io, e.message))
+        } catch (e: IOException) {
+            e.printStackTrace()
+            throw PrintException(context.applicationContext.getString(R.string.err_job_io, e.message))
+        }
+
+        try {
+            future.get(5, TimeUnit.MINUTES)
```

Comment:
Timeout a bit high?

Reply:
That's currently the same as for other printers, I'll keep it for consistency. So far, it has not been an issue, it's just a safeguard for crashed threads